infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java

```diff
@@ -0,0 +1,171 @@
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.schedule.ScheduleDelay;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Periodically checks ability to stop Asynchronous Storage Pod. It will periodically revise
+ * UserPreferences of all registered user and check specialized preferences. Preferences should be
+ * recorded if last workspace stopped and cleanup on start any workspace. Required preferences to
+ * initiate stop procedure for Asynchronous Storage Pod : {@link
+ * org.eclipse.che.api.workspace.shared.Constants#LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE} : should
+ * contain last used infrastructure namespace {@link
+ * org.eclipse.che.api.workspace.shared.Constants#LAST_ACTIVITY_TIME} : seconds then workspace
+ * stopped in the Java epoch time format (aka Unix time)
+ */
+@Singleton
+public class AsyncStoragePodWatcher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStoragePodWatcher.class);
+
+  private final KubernetesClientFactory kubernetesClientFactory;
+  private final UserManager userManager;
+  private final PreferenceManager preferenceManager;
+  private final WorkspaceRuntimes runtimes;
+  private final long shutdownTimeoutSec;
+  private final Clock clock = Clock.systemDefaultZone();
+  private final boolean isAsyncStoragePodCanBeRun;
+
+  @Inject
+  public AsyncStoragePodWatcher(
+      KubernetesClientFactory kubernetesClientFactory,
+      UserManager userManager,
+      PreferenceManager preferenceManager,
+      WorkspaceRuntimes runtimes,
+      @Named("che.infra.kubernetes.async.storage.shutdown_timeout_min") long shutdownTimeoutMin,
+      @Named("che.infra.kubernetes.pvc.strategy") String pvcStrategy,
+      @Named("che.infra.kubernetes.namespace.allow_user_defined")
+          boolean allowUserDefinedNamespaces,
+      @Nullable @Named("che.infra.kubernetes.namespace.default") String defaultNamespaceName,
+      @Named("che.limits.user.workspaces.run.count") int runtimesPerUser) {
+    this.kubernetesClientFactory = kubernetesClientFactory;
+    this.userManager = userManager;
+    this.preferenceManager = preferenceManager;
+    this.runtimes = runtimes;
+    this.shutdownTimeoutSec = TimeUnit.MINUTES.toSeconds(shutdownTimeoutMin);
+
+    isAsyncStoragePodCanBeRun =
+        isAsyncStoragePodCanBeRun(
+            pvcStrategy, allowUserDefinedNamespaces, defaultNamespaceName, runtimesPerUser);
+  }
+
+  /**
+   * Checking current system configuration on ability to run Async Storage Pod. Will be checked next
+   * value of properties:
+   *
+   * <ul>
+   *   <li>che.infra.kubernetes.namespace.default=<username>-che
+   *   <li>che.infra.kubernetes.namespace.allow_user_defined=false
+   *   <li>che.infra.kubernetes.pvc.strategy=common
+   *   <li>che.limits.user.workspaces.run.count=1
+   * </ul>
+   */
+  private boolean isAsyncStoragePodCanBeRun(
```

Comment:
didn't you create `AsyncStorageModeValidator` for this?

Reply:
`AsyncStorageModeValidator implements WorkspaceAttributeValidator ` so its other responsibility i guess and logic pretty simple